Refactor for warnings and display to use templates

rvyuha opened the pull request:
Modified parts.qmd and sets.qmd to utilize templates for display allowing for easier editing.
Added new function to verify and populate display templates found in template-utils.R.
Added templates for warnings found in warning-utils.R.
Modified warnings in parts.qmd and sets.qmd to utilize warning templates.

Warnings have been turned on in all of the qmds to verify template functionality. This should be disabled prior to merging.

rvyuha commented:
Currently missing templating for warnings within functions will be added tomorrow morning.

yulric commented on qmd/sets.qmd:
```diff
@@ -15,12 +15,28 @@ Sets are lists of categories and other parts. For example, `Collection category
+# Display string templates
+header_template <-
+  '## {partLabel} {{ #{partID} }}
+  <a href=\"/parts.html#{partID}\">({partID})</a>. {partDesc} {if (partInstr %!=na% constants$dictionary_missing_value_replacement) partInstr else "" }'
```
I don't think templates should have any logic in them, we should keep them clean with just the variables. This will also help with debugging.

It will also help with refactoring common pieces of coded, for example the code in this and the next template where you replace missing value with ""

yulric commented on qmd/sets.qmd:
```diff
@@ -15,12 +15,28 @@ Sets are lists of categories and other parts. For example, `Collection category
+Status: {status}. First Released: {firstReleased}. \\
+Last updated: {lastUpdated}.'
+
+part_bullet_template <- '**{partLabel}** <a href=\"/parts.html#{partID}\">({partID})</a>. {partDesc} </br>Status: {status}. First released: {firstReleased}. Last updated: {lastUpdated}.'
```
This template is used three times in the code and it looks like the code to get the display inputs is the same. Can we pull it out into a function? Specifically these lines:
1. https://github.com/Big-Life-Lab/PHES-ODM-Doc/blob/warning-refactor/qmd/sets.qmd#L165-L176
2. https://github.com/Big-Life-Lab/PHES-ODM-Doc/blob/warning-refactor/qmd/sets.qmd#L233-L244
3. https://github.com/Big-Life-Lab/PHES-ODM-Doc/blob/warning-refactor/qmd/sets.qmd#L288-L299

yulric commented on R/warning-utils.R:
```diff
@@ -0,0 +1,35 @@
+# Source constants
```
Can you format this file better?
1. Add new lines between the functions
2. Add comments for each function
3. Rename `ivalid_cat_link` to `invalid_cat_link`
4. Can you use better grammer in the warnings messages? Full stops and capitalization for example.

yulric commented on R/template-utils.R:
```diff
@@ -0,0 +1,47 @@
+#' 
+#' @param template string template containg variables to insert surrounded by {}
+#' @param display_variables named list with values to insert where name is equal to name in template
+#' @param ID string containing ID used in warning generation
```
Used when disolaying warnings in case of errors in the input. This should allow the end user to easily know which entity in the dictionary has problematic values.

yulric commented on R/template-utils.R:
```diff
@@ -0,0 +1,47 @@
+  detected_variables <- extract_variables(template)
+  for (single_variable in detected_variables) {
+    if(!(single_variable %in% names(display_variables))){
+      stop(glue::glue('{single_variable} Is missing from the passed display'))
```
This should probably collect all the misssing variables and then stop execution. Otherwise it could get annoying running it one at a time.

yulric commented on R/template-utils.R:
```diff
@@ -0,0 +1,47 @@
+#' 
+#' @return vector of variables used
+extract_variables <- function(template){
+  variable_patter <- "\\{(.*?)\\}"
```
variable_pattern

yulric commented on R/template-utils.R:
```diff
@@ -0,0 +1,47 @@
+  for (detected_variable in all_detected_vars) {
+    detected_variable <- gsub('\\{|\\}', '', detected_variable)
+    return_vars <- append(return_vars, detected_variable)
+  }
```
Return statement?

yulric commented on R/template-utils.R:
```diff
@@ -0,0 +1,47 @@
+#' 
+#' @param template string to check for variables
+#' 
+#' @return vector of variables used
```
vector of strings

yulric commented on R/template-utils.R:
```diff
@@ -0,0 +1,47 @@
+  return_vars <- c()
+  # Strip brackets for easier comparison
+  for (detected_variable in all_detected_vars) {
+    detected_variable <- gsub('\\{|\\}', '', detected_variable)
```
If I understand why you're doing ths, the regex above can be modified to \\{+(.*?)\\}+ to accomplish the same thing.

yulric reviewed (changes requested):
1. All of the strings that you've hardcoded when using the warnings should come from the `partLabel` column in the parts sheet. For example, in this [line](https://github.com/Big-Life-Lab/PHES-ODM-Doc/blob/warning-refactor/qmd/tables.qmd#L150), "part label" should use the `partLabel` column value for the `partLabel` part

yulric commented on R/generate-display.R:
```diff
@@ -0,0 +1,87 @@
+#' @param part_bullet_template string containing display template
+#' 
+#' @return glue object string populated using part_info
+bullet_point_template_population <- function(part_info, partID, part_bullet_template){
```
This is only used in the sets.qmd file right? If so, it should be in that file.

yulric replied:
Can you use a function name that starts as a verb? For examplpe, `generate_bullet_point_display`?

yulric replied:
Does the user need to pass in `partID` here? Can we not extract it from `part_info`?

rvyuha replied:
> Does the user need to pass in `partID` here? Can we not extract it from `part_info`?

PartID is sometimes used outside of the function. So retrieving it within the function rather then passing it as an argument seems to be inefficient. For example for lists and categories the partID is used as key for the list where the output of the function is stored.

yulric commented on R/warning-utils.R:
```diff
@@ -0,0 +1,98 @@
+#' 
+#' Warning for when information is missing for a part.
+#' 
+#' @param column_name string storing the column name. 
```
The parameter descriptions should be made more human readable. For example, `The name of the column with missing information`

yulric replied:
Can you go through all the functions and update the parameter descriptions.

rvyuha replied:
> The parameter descriptions should be made more human readable. For example, `The name of the column with missing information`

In previous PR's you have mentioned I should be including the data type of the parameter. But your current suggestion does not mention data type of the input.

yulric reviewed (changes requested):
I don't think you addressed the issues in this comment, https://github.com/Big-Life-Lab/PHES-ODM-Doc/pull/33#discussion_r1154481141

yulric commented:
> > I don't think you addressed the issues in this comment, [#33 (comment)](https://github.com/Big-Life-Lab/PHES-ODM-Doc/pull/33#discussion_r1154481141)
> 
> I recall us agreeing to not move additional functions to the qmd instead using an R file. So I moved the function to sets.R to avoid cluttering sets.qmd

Sorry, I should have been more specific. I was talking about the two comments after the linked one.

rvyuha commented:
> > > I don't think you addressed the issues in this comment, [#33 (comment)](https://github.com/Big-Life-Lab/PHES-ODM-Doc/pull/33#discussion_r1154481141)
> > 
> > 
> > I recall us agreeing to not move additional functions to the qmd instead using an R file. So I moved the function to sets.R to avoid cluttering sets.qmd
> 
> Sorry, I should have been more specific. I was talking about the two comments after the linked one.
> 
> Edit: Just the function naming comment.
> 
> Edit 2: Can the template be put inside the function? Is there a reason for it to be outside?

I thought it was best to keep the templates inside the qmd as something Doug or Mathew can edit easily. Ideally I believe that the qmd should only contain templates and a function call that handles the rest of the logic.
